## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#8fcd30ee-cd57-44ee-b635-7c424b7a3984) |
<a name="8fcd30ee-cd57-44ee-b635-7c424b7a3984"></a>
### 10683-825b906/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:31229
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d13928a1ef56:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 9 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 1 minutes and 23 seconds
(4/15) Running aws-sdk-ruby tests ... done in 23 seconds
(5/15) Running awscli tests ... done in 2 minutes and 43 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 3 minutes and 34 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 40 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 53 seconds
{
  "args": {
    "bucketName": "minio-go-test-r2igwnvfmc0en6ku",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 1021,
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "Unexpected size",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 10 minutes and 35 seconds
(10/15) Running minio-js tests ... FAILED in 46 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 80,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:682:10) at endReadableNT (_stream_readable.js:1252:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(10/15) Running minio-py tests ... FAILED in 18 minutes and 31 seconds
{
  "name": "minio-py:test_thread_safe",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-52880918-ce55-4f0f-8af6-79b7d65deba5",
    "object_name": "7283734a-506e-4b1b-953c-a659f196a927"
  },
  "message": "Sha-sum mismatch on multi-threaded put and get objects",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 145, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1666, in test_thread_safe\n    raise exceptions[0]\n  File \"/mint/run/core/minio-py/tests.py\", line 1631, in get_object_and_check\n    'Sha-sum mismatch on multi-threaded put and '\nValueError: Sha-sum mismatch on multi-threaded put and get objects\n",
  "duration": 33159
}
(10/15) Running s3cmd tests ... done in 2 minutes and 16 seconds
(11/15) Running s3select tests ... done in 39 seconds
(12/15) Running security tests ... done in 0 seconds

Executed 12 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/bucket-replication.go
+++ b/cmd/bucket-replication.go
@@ -241,7 +241,7 @@ func replicateObject(ctx context.Context, objInfo ObjectInfo, objectAPI ObjectLa
 	for k, v := range putOpts.Header() {
 		headerSize += len(k) + len(v)
 	}
-	r := bandwidth.NewMonitoredReader(ctx, globalBucketMonitor, objInfo.Bucket, objInfo.Name, gr, headerSize, b)
+	r := bandwidth.NewMonitoredReader(ctx, globalBucketMonitor, objInfo.Bucket, objInfo.Name, gr, headerSize, b, target.BandwidthLimit)
 
 	_, err = tgt.PutObject(ctx, dest.Bucket, object, r, size, "", "", putOpts)
 	r.Close()

--- a/cmd/notification.go
+++ b/cmd/notification.go
@@ -1290,14 +1290,14 @@ func sendEvent(args eventArgs) {
 func (sys *NotificationSys) GetBandwidthReports(ctx context.Context, buckets ...string) bandwidth.Report {
 	reports := make([]*bandwidth.Report, len(sys.peerClients))
 	g := errgroup.WithNErrs(len(sys.peerClients))
-	for index, peer := range sys.peerClients {
-		if peer == nil {
+	for index, _ := range sys.peerClients {
+		if sys.peerClients[index] == nil {
 			continue
 		}
 		index := index
 		g.Go(func() error {
 			var err error
-			reports[index], err = peer.MonitorBandwidth(ctx, buckets)
+			reports[index], err = sys.peerClients[index].MonitorBandwidth(ctx, buckets)
 			return err
 		}, index)
 	}

--- a/cmd/notification.go
+++ b/cmd/notification.go
@@ -1290,7 +1290,7 @@ func sendEvent(args eventArgs) {
 func (sys *NotificationSys) GetBandwidthReports(ctx context.Context, buckets ...string) bandwidth.Report {
 	reports := make([]*bandwidth.Report, len(sys.peerClients))
 	g := errgroup.WithNErrs(len(sys.peerClients))
-	for index, _ := range sys.peerClients {
+	for index := range sys.peerClients {
 		if sys.peerClients[index] == nil {
 			continue
 		}
@@ -1321,6 +1321,9 @@ func (sys *NotificationSys) GetBandwidthReports(ctx context.Context, buckets ...
 			if !ok {
 				consolidatedReport.BucketStats[bucket] = bandwidth.Details{}
 				d = consolidatedReport.BucketStats[bucket]
+				d.LimitInBytesPerSecond = report.BucketStats[bucket].LimitInBytesPerSecond
+			}
+			if d.LimitInBytesPerSecond < report.BucketStats[bucket].LimitInBytesPerSecond {
 				d.LimitInBytesPerSecond = report.BucketStats[bucket].LimitInBytesPerSecond
 			}
 			d.CurrentBandwidthInBytesPerSecond += report.BucketStats[bucket].CurrentBandwidthInBytesPerSecond

--- a/pkg/bucket/bandwidth/monitor.go
+++ b/pkg/bucket/bandwidth/monitor.go
@@ -26,16 +26,16 @@ import (
 )
 
 // throttleBandwidth gets the throttle for bucket with the configured value
-func (m *Monitor) throttleBandwidth(ctx context.Context, bucket string, bandwidthBytesPerSecond int64) *throttle {
+func (m *Monitor) throttleBandwidth(ctx context.Context, bucket string, bandwidthBytesPerSecond int64, clusterBandwidth int64) *throttle {
 	m.lock.Lock()
 	defer m.lock.Unlock()
 	throttle, ok := m.bucketThrottle[bucket]
 	if !ok {
-		throttle = newThrottle(ctx, bandwidthBytesPerSecond)
+		throttle = newThrottle(ctx, bandwidthBytesPerSecond, clusterBandwidth)
 		m.bucketThrottle[bucket] = throttle
 		return throttle
 	}
-	throttle.SetBandwidth(bandwidthBytesPerSecond)
+	throttle.SetBandwidth(bandwidthBytesPerSecond, clusterBandwidth)
 	return throttle
 }
 
@@ -124,7 +124,7 @@ func (m *Monitor) getReport(selectBucket SelectionFunction) *bandwidth.Report {
 			continue
 		}
 		report.BucketStats[bucket] = bandwidth.Details{
-			LimitInBytesPerSecond:            m.bucketThrottle[bucket].bytesPerSecond,
+			LimitInBytesPerSecond:            m.bucketThrottle[bucket].clusterBandwidth,
 			CurrentBandwidthInBytesPerSecond: bucketMeasurement.getExpMovingAvgBytesPerSecond(),
 		}
 	}

--- a/pkg/bucket/bandwidth/monitor.go
+++ b/pkg/bucket/bandwidth/monitor.go
@@ -80,7 +80,7 @@ type Monitor struct {
 func NewMonitor(doneCh <-chan struct{}) *Monitor {
 	m := &Monitor{
 		activeBuckets:         make(map[string]*bucketMeasurement),
-		bucketMovingAvgTicker: time.NewTicker(1 * time.Second),
+		bucketMovingAvgTicker: time.NewTicker(2 * time.Second),
 		pubsub:                pubsub.New(),
 		bucketThrottle:        make(map[string]*throttle),
 		doneCh:                doneCh,
@@ -138,7 +138,6 @@ func (m *Monitor) process(doneCh <-chan struct{}) {
 			m.processAvg()
 		case <-doneCh:
 			return
-		default:
 		}
 	}
 }

--- a/pkg/bucket/bandwidth/monitor_test.go
+++ b/pkg/bucket/bandwidth/monitor_test.go
@@ -35,8 +35,8 @@ func TestMonitor_GetThrottle(t *testing.T) {
 		bucket          string
 		bpi             int64
 	}
-	t1 := newThrottle(context.Background(), 100)
-	t2 := newThrottle(context.Background(), 200)
+	t1 := newThrottle(context.Background(), 100, 1024*1024)
+	t2 := newThrottle(context.Background(), 200, 1024*1024)
 	tests := []struct {
 		name   string
 		fields fields
@@ -68,7 +68,7 @@ func TestMonitor_GetThrottle(t *testing.T) {
 			m := &Monitor{
 				bucketThrottle: tt.fields.bucketThrottles,
 			}
-			if got := m.throttleBandwidth(context.Background(), tt.fields.bucket, tt.fields.bpi); got.bytesPerInterval != tt.want.bytesPerInterval {
+			if got := m.throttleBandwidth(context.Background(), tt.fields.bucket, tt.fields.bpi, 1024*1024); got.bytesPerInterval != tt.want.bytesPerInterval {
 				t.Errorf("throttleBandwidth() = %v, want %v", got, tt.want)
 			}
 		})
@@ -135,7 +135,8 @@ func TestMonitor_GetReport(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 			t.Parallel()
 			thr := throttle{
-				bytesPerSecond: 1024 * 1024,
+				bytesPerSecond:   1024 * 1024,
+				clusterBandwidth: 1024 * 1024,
 			}
 			m := &Monitor{
 				activeBuckets:  tt.fields.activeBuckets,

--- a/pkg/bucket/bandwidth/reader.go
+++ b/pkg/bucket/bandwidth/reader.go
@@ -37,7 +37,7 @@ type MonitoredReader struct {
 }
 
 // NewMonitoredReader returns a io.ReadCloser that reports bandwidth details
-func NewMonitoredReader(ctx context.Context, monitor *Monitor, bucket string, object string, reader io.Reader, headerSize int, bandwidthBytesPerSecond int64) *MonitoredReader {
+func NewMonitoredReader(ctx context.Context, monitor *Monitor, bucket string, object string, reader io.Reader, headerSize int, bandwidthBytesPerSecond int64, clusterBandwidth int64) *MonitoredReader {
 	timeNow := time.Now()
 	b := monitor.track(bucket, object, timeNow)
 	return &MonitoredReader{
@@ -47,7 +47,7 @@ func NewMonitoredReader(ctx context.Context, monitor *Monitor, bucket string, ob
 		reader:            reader,
 		lastStop:          timeNow,
 		headerSize:        headerSize,
-		throttle:          monitor.throttleBandwidth(ctx, bucket, bandwidthBytesPerSecond),
+		throttle:          monitor.throttleBandwidth(ctx, bucket, bandwidthBytesPerSecond, clusterBandwidth),
 		monitor:           monitor,
 	}
 }

--- a/pkg/bucket/bandwidth/throttle.go
+++ b/pkg/bucket/bandwidth/throttle.go
@@ -33,21 +33,23 @@ type throttle struct {
 	freeBytes        int64        // unused bytes in the interval
 	bytesPerSecond   int64        // max limit for bandwidth
 	bytesPerInterval int64        // bytes allocated for the interval
+	clusterBandwidth int64        // Cluster wide bandwidth needed for reporting
 	cond             *sync.Cond   // Used to notify waiting threads for bandwidth availability
 }
 
 // newThrottle returns a new bandwidth throttle. Set bytesPerSecond to 0 for no limit
-func newThrottle(ctx context.Context, bytesPerSecond int64) *throttle {
+func newThrottle(ctx context.Context, bytesPerSecond int64, clusterBandwidth int64) *throttle {
 	if bytesPerSecond == 0 {
 		return &throttle{}
 	}
 	t := &throttle{
-		bytesPerSecond: bytesPerSecond,
-		generateTicker: time.NewTicker(throttleInternal),
+		bytesPerSecond:   bytesPerSecond,
+		generateTicker:   time.NewTicker(throttleInternal),
+		clusterBandwidth: clusterBandwidth,
 	}
 
 	t.cond = sync.NewCond(&sync.Mutex{})
-	t.SetBandwidth(bytesPerSecond)
+	t.SetBandwidth(bytesPerSecond, clusterBandwidth)
 	t.freeBytes = t.bytesPerInterval
 	go t.generateBandwidth(ctx)
 	return t
@@ -79,7 +81,7 @@ func (t *throttle) GetLimitForBytes(want int64) int64 {
 }
 
 // SetBandwidth sets a new bandwidth limit in bytes per second.
-func (t *throttle) SetBandwidth(bandwidthBiPS int64) {
+func (t *throttle) SetBandwidth(bandwidthBiPS int64, clusterBandwidth int64) {
 	bpi := int64(throttleInternal) * bandwidthBiPS / int64(time.Second)
 	atomic.StoreInt64(&t.bytesPerInterval, bpi)
 }

--- a/pkg/bucket/bandwidth/throttle.go
+++ b/pkg/bucket/bandwidth/throttle.go
@@ -29,12 +29,14 @@ const (
 
 // throttle implements the throttling for bandwidth
 type throttle struct {
-	generateTicker   *time.Ticker // Ticker to generate available bandwidth
-	freeBytes        int64        // unused bytes in the interval
-	bytesPerSecond   int64        // max limit for bandwidth
-	bytesPerInterval int64        // bytes allocated for the interval
-	clusterBandwidth int64        // Cluster wide bandwidth needed for reporting
-	cond             *sync.Cond   // Used to notify waiting threads for bandwidth availability
+	generateTicker   *time.Ticker    // Ticker to generate available bandwidth
+	freeBytes        int64           // unused bytes in the interval
+	bytesPerSecond   int64           // max limit for bandwidth
+	bytesPerInterval int64           // bytes allocated for the interval
+	clusterBandwidth int64           // Cluster wide bandwidth needed for reporting
+	cond             *sync.Cond      // Used to notify waiting threads for bandwidth availability
+	goGenerate       int64           // Flag to track if generate routine should be running. 0 == stopped
+	ctx              context.Context // Context for generate
 }
 
 // newThrottle returns a new bandwidth throttle. Set bytesPerSecond to 0 for no limit
@@ -46,12 +48,12 @@ func newThrottle(ctx context.Context, bytesPerSecond int64, clusterBandwidth int
 		bytesPerSecond:   bytesPerSecond,
 		generateTicker:   time.NewTicker(throttleInternal),
 		clusterBandwidth: clusterBandwidth,
+		ctx:              ctx,
 	}
 
 	t.cond = sync.NewCond(&sync.Mutex{})
 	t.SetBandwidth(bytesPerSecond, clusterBandwidth)
 	t.freeBytes = t.bytesPerInterval
-	go t.generateBandwidth(ctx)
 	return t
 }
 
@@ -76,6 +78,11 @@ func (t *throttle) GetLimitForBytes(want int64) int64 {
 			}
 		}
 		atomic.AddInt64(&t.freeBytes, -send)
+
+		// Bandwidth was consumed, start generate routine to allocate bandwidth
+		if atomic.CompareAndSwapInt64(&t.goGenerate, 0, 1) {
+			go t.generateBandwidth(t.ctx)
+		}
 		return send
 	}
 }
@@ -96,6 +103,11 @@ func (t *throttle) generateBandwidth(ctx context.Context) {
 	for {
 		select {
 		case <-t.generateTicker.C:
+			if atomic.LoadInt64(&t.freeBytes) == atomic.LoadInt64(&t.bytesPerInterval) {
+				// No bandwidth consumption stop the routine.
+				atomic.StoreInt64(&t.goGenerate, 0)
+				return
+			}
 			// A new window is available
 			t.cond.L.Lock()
 			atomic.StoreInt64(&t.freeBytes, atomic.LoadInt64(&t.bytesPerInterval))
@@ -103,7 +115,6 @@ func (t *throttle) generateBandwidth(ctx context.Context) {
 			t.cond.L.Unlock()
 		case <-ctx.Done():
 			return
-		default:
 		}
 	}
 }

